### ISSUE TITLE
Bad compatibility with https  

### ISSUE THREAD

**Issue by Piterden** [anomalylabs/streams-platform#315]
There is many fixes that I have done to make it work with `https` in admin interface. Almost every of it should work only with https, thats because I can't make PR.

Tell me please can we use relative routing for all parts of system? And if yes - what is the solution?

**Comment by RyanThompson** [comment]
Can you give me an example of some of the issues you are experiencing? Paths should be used now anyways so I am very curious as to what you're running into. Thanks!

**Comment by Piterden** [comment]
First of all I want to say that I have updating CMS on time - almost every day.

I had an issue with incorrect `http` links in `https` environment which I had to fix with some magic JS strings. Here it is:
```js
+(function(document, window, undefined) {

    document.addEventListener('DOMContentLoaded', function() {
        var
            links = document.querySelectorAll('[href^="http:"]'),
            forms = document.querySelectorAll('[action^="http:"]'),
            search = window.location.search,

            fixUrls = function(els, attr) {
                els.forEach(function(el) {
                    var href = el.getAttribute(attr).replace('http:', 'https:');
                    if (attr == 'action' && search !== '') {
                        href += search
                    }
                    return el.setAttribute(attr, href);
                });
            };

        fixUrls(links, 'href');
        fixUrls(forms, 'action');

    });

})(document, window);
```
I put it in initialize.js of PyrocmsTheme (I had to make copy of theme to upgrade the code)

Also I had some problems with incorrect JS constants in `constants.js` in same place.
There are changes of it:
```twig
<script type="text/javascript">

    var APPLICATION_URL = "{{ secure_url('/') }}";
    var APPLICATION_REFERENCE = "{{ env('APPLICATION_REFERENCE') }}";
    var APPLICATION_DOMAIN = "{{ env('APPLICATION_DOMAIN') }}";

    var CSRF_TOKEN = "{{ csrf_token() }}";
    var APP_DEBUG = "{{ config_get('app.debug') }}";
    var APP_URL = "{{ config_get('app.url') }}";
    var REQUEST_ROOT = "{{ request_root()|replace({ 'http': 'https' }) }}";
    var REQUEST_ROOT_PATH = "{{ parse_url(request_root()).path }}";
    var TIMEZONE = "{{ config_get('app.timezone') }}";
    var LOCALE = "{{ config_get('app.locale') }}";
</script>
```
Another issue that I had to fix is the parts of DOM which we get through XHR request and which have `http://` hardcoded inside it. I needed to process this parts on a fly, replacing by regex all `http` to `https` in `modal.js` file:
```js
$(function () {

    var modal = $('.modal');
    var remote = $('.modal.remote');

    // Clear remote modals when closed.
    remote.on('hidden.bs.modal', function () {

        $(this).removeData('bs.modal');

        $(this).find('.modal-content').html('<div class="modal-loading"><div class="active loader"></div></div>');
    });

    // Show loader for remote modals.
    remote.on('show.bs.modal', function () {
        $(this).find('.modal-content').html('<div class="modal-loading"><div class="active loader"></div></div>');
    });

    // Handle ajax links in modals.
    modal.on('click', 'a.ajax, .pagination a', function (e) {

        e.preventDefault();

        var wrapper = $(this).closest('.modal-content');

        wrapper.append('<div class="modal-loading"><div class="active loader"></div></div>');

        $.get($(this).attr('href').replace('http:', 'https:'), function (html) {
            wrapper.html(html.replace(new RegExp(/http:/, 'g'), 'https:'));
        });
    });

    // Handle ajax forms in modals.
    modal.on('submit', 'form.ajax', function (e) {

        e.preventDefault();

        var wrapper = $(this).closest('.modal-content');

        wrapper.append('<div class="modal-loading"><div class="active loader"></div></div>');

        if ($(this).attr('method') == 'GET') {
            $.get($(this).attr('action').replace('http:', 'https:'), $(this).serializeArray(), function (html) {
                wrapper.html(html.replace(new RegExp(/http:/, 'g'), 'https:'));
            });
        } else {
            $.post($(this).attr('action').replace('http:', 'https:'), $(this).serializeArray(), function (html) {
                wrapper.html(html.replace(new RegExp(/http:/, 'g'), 'https:'));
            });
        }
    });
});

```

**Comment by RyanThompson** [comment]
Which modals exactly were misbehaving? There is really no need to do this JS magic - we need to get a look at the source / origin and find out why it's including protocol at all for one but also why it seems to be cached that way and not reflecting http. Can you point me in the direction to replicate these issues? Which modals exactly and such.

**Comment by Piterden** [comment]
Obviously, there is no need for this magic, but things don't work like I expect without it. Not only modals, there is many other things.
Have you tried to install Pyro on secure host yourself?

You see, now I don't have enough time to describe all this issues. Maybe we could do it otherwise?
I can give you access to admin panel and disable all my magic for some time for you.

If this way is OK for you, please tell me and let do it.

**Comment by RyanThompson** [comment]
Ya send me the credentials to login via email if you like I'll check it out. We use SSL all the time so this is weird lol

**Comment by Piterden** [comment]
I have send you email with cred-s to mailbox from your profile here

**Comment by Piterden** [comment]
Hi, Ryan!

I have lost you. Please let me know if you have watch at our system or not.
Because I could continue development only after your answer.

Thanks.

**Comment by Piterden** [comment]
Hello, Ryan!

I want to tell you, that I can keep our developing frozen maximum to Monday (2016.11.14) morning. I hope you have time to watch system & understand problems.

**Comment by RyanThompson** [comment]
Hey there - I've been pretty tied up. I will try and get to it today or tomorrow but we work with SSL all the time. Have _you_ had any time to look into it yourself again?

**Comment by Piterden** [comment]
Yes, I was able to make SSL work, but only on new fresh system. This is not Nginx for now, because servers have same configs. I suppose it is LetsEncrypt certificate problem or maybe something else...

Anyway, the great thank to you, for spend your time!

Best solution for me should be using fresh system with working SSL from the box and transporting my addons to that system.
And also thanks to the Pyro and Streams platform, based on their modular principles I can easily find other-way-solution to many difficult situations.

**Comment by RyanThompson** [comment]
Ok thank goodness you found the solution! I was pretty certain it was something configuration related but I am glad either way that it's solved! I appreciate the kind words - I'll be seeing you around ^_^
